Writing first spec

edwaltz89 opened the pull request:
Writing first spec about recommendation api
and refer to below ticket as standard
https://github.com/Porters/matching-pro-business/issues/327

I'll be pleased with your feedback or response. thx!

ChihyungPark commented on openapi-assist.yml:
```diff
@@ -0,0 +1,162 @@
+                type: string
+              jobDetail:
+                type: string
+      RecommendationsOutput:
```
everything seems okay, 
but here the output of recommendation seems different from the type in content. 
If it is okay, I'm okay

edwaltz89 replied:
Maybe It's okay. because as feature spec in ticket, pagination with button click is just call 1 record.
so I think It is okay if we don't have to wrapping with array
I'll fix the put method's RecommendationsInput format and re-request review. thx 4 your comment!

SeraphicRav replied:
I think we should use the same input/output format to make typing and manipulation easier. You can reuse schemas in OpenAPI too.

edwaltz89 commented on openapi-assist.yml:
```diff
@@ -0,0 +1,162 @@
+
+components:
+  schemas:
+      RecommedationsInput:
```
@ChihyungPark 
Fixed input schema. thx!

SeraphicRav commented on openapi-assist.yml:
```diff
@@ -0,0 +1,167 @@
+                $ref: '#/components/schemas/InvalidParams'
+        500:
+          description: Internal Error
+    patch:
```
Patching usually asks for a path parameter element representing the id of the updated thing
Maybe having the path /recommendations/{sourceId}/resume/{resumeId}/job/{jobId}

With
```json
{
  "status": "boolean"
}
```

Would be more restlike

edwaltz89 replied:
@SeraphicRav 
Changed parameter from query to path and remove query string in patch method

SeraphicRav reviewed (changes requested):
Questions:
- there is concept of index which only exists in the get endpoint. How do you know the current index ?
- what happens when recommendations get scouted or is marked as not-good ? The index might skip some recommendations

edwaltz89 commented:
> Questions:
> 
> * there is concept of index which only exists in the get endpoint. How do you know the current index ?
> * what happens when recommendations get scouted or is marked as not-good ? The index might skip some recommendations

Answers:

1. there is concept of index which only exists in the get endpoint. How do you know the current index ?
I think that recommendation-api don't have to know current index.
I just expect to receive index from web in order by clicking button and api is just response by index record

2. what happens when recommendations get scouted or is marked as not-good ? The index might skip some recommendations
When recommendation is scouted or marked not good, I expect the status flag is changed and by changed flag, recommendation changed flag is not listed

edwaltz89 commented:
Added fix details after last reviews:
1. Update name and schema of recommendation item
2. Changed parameter from query to path and remove query string in patch method
3. Add the index field in GET method response

SeraphicRav commented on openapi-assist.yml:
```diff
@@ -0,0 +1,168 @@
+            application/json:
+              schema:
+                $ref: '#/components/schemas/InvalidParams'
+        500:
```
Please add 404 in the case that there is no recommendation left

SeraphicRav commented on openapi-assist.yml:
```diff
@@ -0,0 +1,168 @@
+      type: object
+      properties:
+        sourceId:
+          type: number
```
Please add an index

SeraphicRav commented on openapi-assist.yml:
```diff
@@ -0,0 +1,168 @@
+    index:
+      name: index
+      in: query
+      required: true
```
It should be optional.

When doing the first call, the API user does not have any idea about what is the latest index. So when it is not provided, the latest recommendations should be provided

Otherwise this number should be the current position

SeraphicRav commented on openapi-assist.yml:
```diff
@@ -0,0 +1,168 @@
+      required: true
+      description: Index to recommendation pagination
+      schema:
+        type: number
```
Let us add a new query parameter: direction which can be either next or previous so that we know if we need to look for a value smaller or bigger than index

SeraphicRav commented on openapi-assist.yml:
```diff
@@ -0,0 +1,168 @@
+      schema:
+        type: number
+paths:
+  /recommendation:
```
Let us use recommendationS for the path as we are using plural in core and web by default

SeraphicRav commented on openapi-assist.yml:
```diff
@@ -0,0 +1,168 @@
+      type: object
+      properties:
+        statusFlag:
+          type: number
```
I think a string status would be more readable. Let us also name it status to make it smaller

edwaltz89 commented:
Added fix details after last reviews:

Update index parameter to optional
Add direction parameter
Update name in path from recommendation to recommendations

SeraphicRav commented on openapi-assist.yml:
```diff
@@ -0,0 +1,180 @@
+    StatusPatchBody:
+      type: object
+      properties:
+        status:
```
Let us use string for status

SeraphicRav commented on openapi-assist.yml:
```diff
@@ -0,0 +1,180 @@
+      name: direction
+      in: query
+      required: true
+      description: page direction ( prev | next )
```
Let us use the full world previous

SeraphicRav replied:
You can use an enum to limit the possible values

SeraphicRav commented on openapi-assist.yml:
```diff
@@ -0,0 +1,180 @@
+      name: index
+      in: query
+      required: false
+      description: Current index to recommendation pagination (Using with pageDirection)
```
pageDirection became direction

SeraphicRav commented on openapi-assist.yml:
```diff
@@ -0,0 +1,180 @@
+    direction:
+      name: direction
+      in: query
+      required: true
```
direction should not be required, it should be required when you provide an index (Don't know if OpenAPI allows this kind of dependency)
